### PR TITLE
feat: banners

### PR DESCRIPTION
**Pull request by raymond-le-goldenowl**
**Task**

Endpoints of `banners` entity

**Descriptions**

- Create A Banner

  > payload

```json
{
  "icon": "image path",
  "href": "string",
  "icon": "file"
}
```

> response

```json
{
  "id": "uuid",
  "text": "href|url",
  "icon": "image path",
  "href": "string",
  "created_at": "datetime",
  "updated_at": "datetime",
  "deleted_at": "datetime"
}
```

- Get All Banners
  > payload

```json

```

> response

```json
[
  {
    "id": "uuid",
    "text": "href|url",
    "icon": "image path",
    "href": "string",
    "created_at": "datetime",
    "updated_at": "datetime",
    "deleted_at": "datetime"
  },
  {
    "id": "uuid",
    "text": "href|url",
    "icon": "image path",
    "href": "string",
    "created_at": "datetime",
    "updated_at": "datetime",
    "deleted_at": "datetime"
  }
]
```

- Get One Banner

  > payload

  Params:

  ```json
  { "id": "string" }
  ```

> response

```json
{
  "id": "uuid",
  "text": "href|url",
  "icon": "image path",
  "href": "string",
  "created_at": "datetime",
  "updated_at": "datetime",
  "deleted_at": "datetime"
}
```

- Update One Banner

  > payload

  Params:

  ```json
  { "id": "string" }
  ```

  body:

  ```json
  {
    "icon": "image path",
    "href": "string",
    "icon": "file"
  }
  ```

> response

```json
{
  "id": "uuid",
  "text": "href|url",
  "icon": "image path",
  "href": "string",
  "created_at": "datetime",
  "updated_at": "datetime",
  "deleted_at": "datetime"
}
```

- Delete Banners

  > payload

  ```json

  ```

> response

```json

```

- Delete A Banner

  > payload

  Params:

  ```json
  { "id": "string" }
  ```

  Query:

  ```json
  { "remove": "true|false" }
  ```

> response

```json

```

- Delete Banners

  > payload

  ```json

  ```

> response

```json

```

- Get Banners Images

  > payload

  ```json

  ```

> response

response a image

- **Screenshots**

>

### CODE REVIEW

File: src/banners/banners.service.ts [hunk 1]
```diff
@@ -10,23 +18,123 @@ export class BannersService {
...
   }
 
-  remove(id: number) {
-    return `This action removes a #${id} banner`;
+  async deleteAll() {
+    const deletedAll = await this.bannersRepository.delete({});
+
+    if (deletedAll.affected < 0) {
+      throw new BadRequestException(`Can not delete`);
+    }
```
Comment (jun-nguyen-goldenowl) [inline src/banners/banners.service.ts]: when affected < 0?

File: src/banners/banners.controller.ts [hunk 2]
```diff
@@ -16,28 +24,84 @@ import { UpdateBannerDto } from './dto/update-banner.dto';
 export class BannersController {
   constructor(private readonly bannersService: BannersService) {}
```
Comment (jun-nguyen-goldenowl) [inline src/banners/banners.controller.ts:25]: Should apply admin role guard for this controller

File: src/banners/banners.service.ts [hunk 3]
```diff
@@ -10,23 +18,123 @@ export class BannersService {
     @InjectRepository(BannersRepository)
     private bannersRepository: BannersRepository,
   ) {}
-  create(createBannerDto: CreateBannerDto) {
-    return this.bannersRepository.create({});
+
+  async create(icon, createBannerDto: CreateBannerDto): Promise<Banner> {
+    // return icon;
```
Comment (jun-nguyen-goldenowl) [inline src/banners/banners.service.ts]: remove redundant/commented code.

File: src/banners/banners.service.ts [hunk 4]
```diff
@@ -10,23 +18,123 @@ export class BannersService {
...
+    // create path point to images resource
+    const path = `/banners/images/${icon?.filename}`;
+    createBannerDto.icon = path;
+
+    // save and assign a result saved
+    const saved = await this.bannersRepository.save(createBannerDto);
+
+    // if saved fail should be throw new error
+    if (!saved) {
+      throw new BadRequestException(`Can not save object`);
```
Comment (jun-nguyen-goldenowl) [inline src/banners/banners.service.ts]: please give a correct/more detailed error message. As a user, I don't know what `object` is

File: src/banners/banners.service.ts [hunk 5]
```diff
@@ -10,23 +18,123 @@ export class BannersService {
...
+
+    createBannerDto.text = 'search';
+
+    if (isURL(createBannerDto.href)) {
+      createBannerDto.text = 'url';
+    }
+
+    // create path point to images resource
+    const path = `/banners/images/${icon?.filename}`;
+    createBannerDto.icon = path;
```
Comment (jun-nguyen-goldenowl) [inline src/banners/banners.service.ts]: just assign value of variable `path` to `createBannerDto.icon`. No need to declare a variable for it if that variable won't be re-used in anywhere

File: src/banners/banners.service.ts [hunk 6]
```diff
@@ -10,23 +18,123 @@ export class BannersService {
...
-  findOne(id: number) {
-    return `This action returns a #${id} banner`;
+  async findOneBanner(id: string): Promise<Banner> {
+    // find one banner by id and with deleted
+    const banner = await this.bannersRepository.findOne({
+      where: {
+        id,
+      },
+      withDeleted: true,
+    });
```
Comment (jun-nguyen-goldenowl) [inline src/banners/banners.service.ts]: you want to support finding a soft-deleted record. Right?
I think you should also provide an options param to this method for purpose of finding record with/without soft-deleted.

**Review by jun-nguyen-goldenowl — CHANGES_REQUESTED** [review]

File: src/banners/banners.controller.ts (hunk 2 above)
Comment (raymond-le-goldenowl) [reply src/banners/banners.controller.ts:25]: i will update later

**Review by jun-nguyen-goldenowl — APPROVED** [review]

**Comment by jun-nguyen-goldenowl** [comment]
I wil merge this PR into dev to continue on reviewing PR #3

**Pull request merged**